### ISSUE TITLE
Reuest to have one shared observable for all subscriptions

### ISSUE THREAD

**Issue by josvos** [shamblett/mqtt_client#23]
Hi,

It should be possible (and probably even the default, or the only way ;-)) to have *one* observable for all subscriptions, so that all received messages are processed (once) via that single observable.  Now overlapping subscriptions (due to wildcards, for example) may cause the message to notify multiple observables.  Furthermore, due to the topic passed to the observable, there is no need to have more than one observable and it is for most applications a waste of resources (although I admit there might be use cases for it).

None of the MQTT implementations I know do "copy" received messages matching multiple subscription topics.  This is against the concept of how MQTT subscription topics work and how a broker deals with them (it combines all subscriptions), so IMHO it's also bad client application behavior to rely on such a feature.

In case you want to dynamically (un)subscribe and potentially have overlapping topics, you have to filter out the message duplicates now.  A workaround could be to exit the loop in `publishMessageReceived()` after the first match (set by an option, like `notifyOnlyOnce`), assuming all observables share the same listen-function, but this is not very elegant.

**Comment by shamblett** [comment]
OK, I guess we could have just one observable per client that users listen on, we could then leave everything as it is and just update publishMessageReceived() so that on first match of any of our subscriptions we publish to this one observeable, this behaviour could be the default, optionally we could also do as we do now in case anyone still wants this. Would this be sufficient?

**Comment by josvos** [comment]
Not sure what you mean with "we could then leave everything as it is", as `createNewSubscription` now creates a new observable per subscription and that should be taken out: the (one) observable can just be an attribute of the `MqttClient`.

The matching code `SubscriptionTopic.matches()` is now redundant, as all received messages per definition belong to a subscription and are all passed to the same observable.  Checking topics then is a waste of CPU cycles and error sensitive, so I would suggest to remove that completely.

If you want backwards compatibility, it might become a bit messy, maybe isolate that code and let the user enable it with some special option.

**Comment by shamblett** [comment]
Yes, but we could just leave it, the user can use the returned observable from the subscription or not, depending on his choice of usage. Your right in that we don't need to match incoming topics at all now, we can just pass it to the client observable, this way we can just enable/disable the single observable, we don't need to isolate any code, just not use it.

I'm not worried about CPU, the Dart VM is blazingly fast, a few string comparisons aren't going to hurt it.

So we could introduce an option named singleObservable say, add a client method to listen to this, update subscribe to return null maybe or just leave it and comment the API accordingly, internally we still use the observable list but update publishMessageReceived() to not perform any matches and just update the client observable. This way we don't have to disturb any existing code at all.

The reason the client works like this at all is because the C# nMQTT lib its ported from did, its not really something I've thought about to in purely MQTT terms TBH.

**Comment by josvos** [comment]
Well, I always keep worrying about CPU cycles (and other resources), as I'm using it in Flutter apps (in fact, I'm using @benbehringer's fork now, until your code will be Flutter / Dart 2 compatible ;-)).

I would prefer not creating unused observables (not sure if you meant that in your first sentence), again to avoid wasting resources.  Every `subscribe()` could also return that single observable (in case you have chosen to use a single observable), but that is maybe too confusing.

Hope the C# nMQTT lib is not having more strange things...

**Comment by shamblett** [comment]
Ok, updated the code for this change, this wasn't as intrusive as I first thought, works quite well also. I've removed the observables list as you suggest, anyone who wants backwards compatibility could now of course run more than one client if they need to. Lets see if anyone shouts about this.

Re-published at 2.0.0.

**Comment by josvos** [comment]
Sounds good, I didn't test it, as I can only use @benbehringer's fork now.  I'll open a new issue (request) for Dart 2.0.
